merge_values.go

```diff
@@ -0,0 +1,57 @@
+package helmclient
+
+import (
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// yamlToStringMap unmarshals the YAML input into a map[string]interface{}
+// with string keys. This is necessary because the default behaviour of the
+// YAML parser is to return map[interface{}]interface{} types.
+// See https://github.com/go-yaml/yaml/issues/139.
+//
+func yamlToStringMap(input []byte) (map[string]interface{}, error) {
+	var raw interface{}
+	var result map[string]interface{}
+
+	err := yaml.Unmarshal(input, &raw)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	output := processMapValue(raw)
+	result = output.(map[string]interface{})
```

Comment:
I don't think it's correct to always assert processMapValue output type as `map[string]interface{}`. Why it wouldn't return that type instead?

Reply:
I want to base the merge logic on the upstream logic in Helm. It does what we need and we already use it. Albeit indirectly because we shell out to Helm in draughtsman.

https://github.com/helm/helm/blob/240e539cec44e2b746b3541529d41f4ba01e77df/cmd/helm/install.go#L358

For the merge logic to work it needs string keys. This is so it can identify nested values and do the recursive merge.

Reply:
Sure, what I'm trying to say here this type assertion will fail at runtime when input will be of type string for example.

Unless there is an assumption that the input is always an object. But then I think it should be clear to avoid confusion. Like I have right now. This could be achieved by extracting `processValue` from `processMapValue`. Then `processMapValue` can return the desired type.

Reply:
OK I think I get you. Extracting `processValue` is a good shout. I'll look into that.

Reply:
What I'd like to do here is something like:

```go
inputMap, ok := raw.(map[interface{}]interface{})
if !ok {
	return nil, return microerror.Maskf(executionFailedError, "input type %T but expected %T", raw, inputMap)
}

result = processInterfaceMap(raw)
```

This would make clear what is the expected input type.

Other question: why we can't just do following in the first place?

```go
var raw map[string]interface{}

err := yaml.Unmarshal(input, &raw)
```

Reply:
That is nice to check the input type. Done.

The problem is the nested keys will be `interface{}` and not `string`. In V3 of `go-yaml` they want to change the behaviour but there is no date for V3. Until then this workaround is needed.

Reply:
Ack 👍 Maybe a comment in the code?

Reply:
This is already in the godoc and it links to the upstream issue. Going with this.